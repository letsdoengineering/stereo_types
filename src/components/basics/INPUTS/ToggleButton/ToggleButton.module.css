/* Hide default HTML checkbox */
.container input {
    margin: 0;
    opacity: 0;
    width: 0;
    height: 0;
}

/* Container to fill its parent */
.container {
    display: flex;
    height: 100%;
}

/* Text to sit center within the toggle */
.text {
    margin: auto;
}

/* Un-checked style of toggle button */
.toggle {
    display: flex;
    width: 100%;
    padding: 1rem;
    cursor: pointer;
    background: white;
    border: 1px solid var(--outline-gray);
}

/* Checked style */
.container input:checked + .toggle {
    background: var(--highlight-blue-20);
    color: var(--dark-gray);
    border: 1px solid var(--outline-gray);
}

/* focus style */
.container input:focus + .toggle {
    -moz-box-shadow:    0 0 0 2px var(--focus);
    -webkit-box-shadow: 0 0 0 2px var(--focus);
    box-shadow:         0 0 0 2px var(--focus);
}
