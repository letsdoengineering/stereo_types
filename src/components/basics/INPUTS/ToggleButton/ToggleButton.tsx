import React from 'react'
import Text from './../../Text/Text'
import * as styles from './ToggleButton.module.css'

type ToggleButtonProps = {
  /** Tell the button (checkbox) it's status; controlled by consumer via passing onChange function */
  isChecked: boolean
  /** Called when clicked */
  onChange(): void
  /** Text label */
  label: string
}

const ToggleButton: React.FC<ToggleButtonProps> = ({
  isChecked,
  onChange,
  label,
  ...rest
}: ToggleButtonProps) => (
  <label className={styles.container}>
    <input type='checkbox' checked={isChecked} onChange={onChange} {...rest} />
    <div className={styles.toggle}>
      <Text className={styles.text}>{label}</Text>
    </div>
  </label>
)

export default ToggleButton
