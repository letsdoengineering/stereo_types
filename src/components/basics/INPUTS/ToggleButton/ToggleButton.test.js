import * as React from 'react'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'

import ToggleButton from './ToggleButton'

describe('ToggleButton', () => {
    it('should call onChange when clicked', () => {
        const onChange = jest.fn()
        render(<ToggleButton isChecked={false} onChange={onChange} label={'a label'} />)
        const toggle = screen.getByLabelText('a label')
        userEvent.click(toggle)
        expect(onChange).toHaveBeenCalled()
    })

    it('should render correctly', () => {
        const onChange = jest.fn()
        const tree = renderer
            .create(<ToggleButton isChecked={true} onChange={onChange} label={'a label'} />)
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const onChange = jest.fn()
        const { container } = render(
            <ToggleButton isChecked={false} onChange={onChange} label={'a label'} />
        )
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
