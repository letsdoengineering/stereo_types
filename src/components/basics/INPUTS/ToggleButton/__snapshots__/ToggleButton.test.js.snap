// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton should render correctly 1`] = `
<label
  className="container"
>
  <input
    checked={true}
    onChange={[MockFunction]}
    type="checkbox"
  />
  <div
    className="toggle"
  >
    <span
      className="sizeS weightNormal textGray text"
    >
      a label
    </span>
  </div>
</label>
`;
